### PR TITLE
verify tx with ledger when proposing a block

### PR DESCRIPTION
**Pull request by trueinsider**
### Proposed changes in this pull request
There is times when tx (already invalid) getting added to a proposed block. Block proposer should verify such txs and don't add them to the block.

### Type (put an `x` where ever applicable)
- [ ] Bug fix: Link to the issue
- [x] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [ ] Documentation Improvement

### Checklist
Please put an `x` against the checkboxes. Write a small comment explaining if its `N/A` (not applicable)

- [x] Read the [CONTRIBUTION guidelines](https://github.com/nknorg/nkn#contributing).
- [x] All the tests are passing after the introduction of new changes.
- [ ] Added tests respective to the part of code I have written.
- [ ] Added proper documentation where ever applicable (in code and README.md).
- [x] Code has been written according to [NKN-Golang-Style-Guide](https://github.com/nknorg/nkn/wiki/NKN-Golang-Style-Guide)

### Extra information
Any extra information related to this pull request.

### CODE REVIEW

**Comment by yilunzhang** [comment]
Since this is a compatible change, we will test and merge it after mainnet launch :)

**Comment by NoelBright** [comment]
Looks good to me.

File: chain/txValidator.go [hunk 1]
```diff
@@ -488,15 +487,13 @@ func (bvs *BlockValidationState) VerifyTransactionWithBlock(txn *transaction.Tra
...
-		if header != nil {
-			coinbase := payload.(*pb.Coinbase)
-			donationAmount, err := DefaultLedger.Store.GetDonation()
-			if err != nil {
-				return err
-			}
-			if Fixed64(coinbase.Amount) != GetRewardByHeight(header.UnsignedHeader.Height)+donationAmount {
-				return errors.New("Mining reward incorrectly.")
-			}
+		coinbase := payload.(*pb.Coinbase)
```
Comment (yilunzhang) [inline chain/txValidator.go]: Why not skipping the check when height is 0?

File: chain/mining.go [hunk 2]
```diff
@@ -93,17 +95,30 @@ func (bm *BuiltinMining) BuildBlock(ctx context.Context, height uint32, chordID
...
+			continue
+		}
+		if err := VerifyTransactionWithLedger(txn); err != nil {
+			log.Warning("invalid transaction")
+			txnCollection.Pop()
+			continue
+		}
+		if err := bvs.VerifyTransactionWithBlock(txn, height); err != nil {
+			bvs.Reset()
+			log.Warning("invalid transaction")
```
Comment (yilunzhang) [inline chain/mining.go]: It's better to use `log.Warningf("invalid transaction: %v", err)` here and in L104

File: chain/txValidator.go (hunk 1 above)
Comment (trueinsider) [reply chain/txValidator.go]: height is only 0 when called from txpool and txpool don't have coinbase txs

Comment (yilunzhang) [reply chain/txValidator.go]: I think we should further make sure that by adding a switch condition in pool.processTx that returns error when type is coinbase

**Pull request merged**